icu4c/source/common/cmemory.h

```diff
@@ -37,19 +37,63 @@
 #include <stdio.h>
 #endif
 
-
+// uprv_memcpy
+#if defined(__clang__)
+#define uprv_memcpy(dst, src, size) UPRV_BLOCK_MACRO_BEGIN { \
+    /* Suppress warnings about addresses that will never be NULL */ \
+    _Pragma("clang diagnostic push") \
+    _Pragma("clang diagnostic ignored \"-Waddress\"") \
+    U_ASSERT(dst != NULL); \
+    U_ASSERT(src != NULL); \
+    U_STANDARD_CPP_NAMESPACE memcpy(dst, src, size); \
+    _Pragma("clang diagnostic pop") \
+} UPRV_BLOCK_MACRO_END
+#elif defined(__GNUC__)
 #define uprv_memcpy(dst, src, size) UPRV_BLOCK_MACRO_BEGIN { \
+    /* Suppress warnings about addresses that will never be NULL */ \
+    _Pragma("GCC diagnostic push") \
+    _Pragma("GCC diagnostic ignored \"-Waddress\"") \
     U_ASSERT(dst != NULL); \
     U_ASSERT(src != NULL); \
     U_STANDARD_CPP_NAMESPACE memcpy(dst, src, size); \
+    _Pragma("GCC diagnostic pop") \
 } UPRV_BLOCK_MACRO_END
+#else
+#define uprv_memcpy(dst, src, size) UPRV_BLOCK_MACRO_BEGIN { \
+    U_ASSERT(dst != NULL); \
+    U_ASSERT(src != NULL); \
+    U_STANDARD_CPP_NAMESPACE memcpy(dst, src, size); \
+} UPRV_BLOCK_MACRO_END
+#endif
 
+// uprv_memmove
+#if defined(__clang__)
```

Comment:
optional: maybe use one single if-elif chain, with both memcpy and memmove next to each other in each branch?

Reply:
good idea, I think that would help to clean it up. :)